### PR TITLE
Add backend telemetry events in ML

### PR DESCRIPTION
**Pull request by petersg83**
### What does it do?

Sends some telemetry information about the Media Library (mainly folders) once a week

### Why is it needed?

To understand how the folders are used and to see if the feature can be improved for specific use-cases

### CODE REVIEW

File: packages/core/upload/server/services/metrics.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+      const pingCron = scheduleJob(getCronRandomHour(), async () => {
+        const metrics = await this.computeMetrics();
+        strapi.telemetry.send('mediaLibraryMetrics', metrics);
+      });
+
+      crons.push(pingCron);
+    },
+
+    destroy() {
```
Comment (gu-stav) [inline packages/core/upload/server/services/metrics.js:82]: I'm not sure if we are using `destroy` as a convention, but for this case I find it a bit harsh. What do you think about reversing the initialization, e.g. `stopRegularMetricsUpdate()`?

File: packages/core/upload/server/services/metrics.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          maxDepth = folderLevel.depth;
+        }
+      }
+      const averageDepth = product / folderNumber;
+
+      let sumOfDeviation = 0;
+      for (const folderLevel of folderLevelsArray) {
+        sumOfDeviation += Math.abs(folderLevel.depth * folderLevel.occurence - averageDepth);
+      }
+      const averageDeviationDepth = sumOfDeviation / folderNumber;
```
Comment (gu-stav) [inline packages/core/upload/server/services/metrics.js:54]: I think it would be great to have a test for this part of the code, as it is hard to follow and might break easily. Do you think you could extract this into a function and make it testable?

File: packages/core/upload/server/services/metrics.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
+'use strict';
+
+const { scheduleJob } = require('node-schedule');
+const { FOLDER_MODEL_UID, FILE_MODEL_UID } = require('../constants');
+
+const rand = max => Math.floor(Math.random() * max);
+const getCronRandomHour = () => `${rand(60)} ${rand(60)} ${rand(24)} ${rand(7)} * *`;
```
Comment (gu-stav) [inline packages/core/upload/server/services/metrics.js:7]: Out of curiosity: what is the benefit of randomizing this? To prevent many requests at the same time from different instances?

File: packages/core/upload/server/services/metrics.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      const knex = strapi.db.connection;
+      const folderLevelsArray = (
+        await knex(folderTable)
+          .select(
+            knex.raw(
+              `LENGTH(${keepOnlySlashesSQLString}) as depth, count(*) as occurence`,
+              queryParams
+            )
+          )
+          .groupBy('depth')
```
Comment (gu-stav) [inline packages/core/upload/server/services/metrics.js:35]: Could you try and explain me what this part of the code does?

File: packages/core/upload/server/controllers/admin-folder-file.js [hunk 5]
```diff
@@ -32,7 +32,20 @@ module.exports = {
...
 
     const deletedFiles = await fileService.deleteByIds(body.fileIds);
-    const deletedFolders = await folderService.deleteByIds(body.folderIds);
+    const {
+      folders: deletedFolders,
+      totalFolderNumber,
+      totalFileNumber,
+    } = await folderService.deleteByIds(body.folderIds);
+
+    if (deletedFiles.length + deletedFolders.length > 1) {
```
Comment (gu-stav) [inline packages/core/upload/server/controllers/admin-folder-file.js:41]: Shouldn't this be `>= 1`?

File: packages/core/upload/server/controllers/admin-folder-file.js [hunk 6]
```diff
@@ -152,7 +168,7 @@ module.exports = {
             );
 
           // update path of files below
-          await strapi.db
+          totalFileNumber = await strapi.db
```
Comment (gu-stav) [inline packages/core/upload/server/controllers/admin-folder-file.js:171]: Doesn't `.update()` return an array of updated entries?

File: packages/core/upload/server/services/metrics.js (hunk 1 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:82]: Yes it's better I agree!

File: packages/core/upload/server/services/metrics.js (hunk 2 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:54]: I was thinking about that too but I think remembering that we don't do tests on the telemetry. @alexandrebodin any policy regarding that?

File: packages/core/upload/server/services/metrics.js (hunk 3 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js]: Yes exactly :)

File: packages/core/upload/server/services/metrics.js (hunk 4 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:35]: Yes :)
In order to get the depth level of a folder, I take their path and remove all numbers and count (using `LENGTH`) the remaining `/` (I couldn't find a better way to do it than nested `REPLACE`).
I have a row per folder, with their depth. Then, by grouping by their depth and using `count(*)`, it gives me, for each depth, the number of folders that have this depth.

| depth  | occurence |
| ------ | ----------- |
| 1  | 4  |
| 2  | 2 |
| 3  | 5 |

In the end the query returns:
```
[
  { depth: 1, occurence: 4 },
  { depth: 2, occurence: 2 },
  { depth: 3, occurence: 5 },
]
```

Do you think I should write all of this in a comment?

File: packages/core/upload/server/controllers/admin-folder-file.js (hunk 5 above)
Comment (petersg83) [reply packages/core/upload/server/controllers/admin-folder-file.js:41]: No because otherwise we don't know if the user clicked on the bulk button or not (because this route is also used when you delete a folder through the edit view of the folder), which is the thing we want to track

File: packages/core/upload/server/controllers/admin-folder-file.js (hunk 6 above)
Comment (petersg83) [reply packages/core/upload/server/controllers/admin-folder-file.js:171]: Only if we ask it to, by using `.returning()` or giving returning params in `.update()` :)

File: packages/core/upload/server/services/metrics.js (hunk 3 above)
Comment (gu-stav) [reply packages/core/upload/server/services/metrics.js]: You thought of everything, nice!

File: packages/core/upload/server/services/metrics.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      };
+    },
+
+    async startRegularMetricsUpdate() {
+      if (running) {
+        throw new Error('Regular metrics updates are already running');
+      }
+      running = true;
+
+      const pingCron = scheduleJob(getCronRandomWeekly(), async () => {
```
Comment (alexandrebodin) [inline packages/core/upload/server/services/metrics.js:74]: Wy not use the cron service to register the jobs instead of installing node-schedule as a dependency here?

Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:75]: Because the user can disable the cron service. I think it's why we also don't use the cron service in core metrics too.
Would you create a parallel cron service for internal use?

File: packages/core/upload/server/services/metrics.js (hunk 4 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:35]: The SQL query looks like that:
```
'LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(REPLACE(??, ?, ?), ?, ?), ?, ?), ?, ?), ?, ?), ?, ?), ?, ?), ?, ?), ?, ?), ?, ?)) as depth, count(*) as occurence'
```
It's because the `REPLACE` function cannot take a regex, only a string. I want to remove all numbers so I call the `REPLACE` function 10 times, one for each number

File: packages/core/upload/server/services/metrics.js (hunk 7 above)
Comment (alexandrebodin) [reply packages/core/upload/server/services/metrics.js:75]: Hum 🤔  I would suggest we refactor this so the cron service just works and if the user disables the cron config it just doesn't register the user crons instead. What do you think ?

File: package.json [hunk 8]
```diff
@@ -98,9 +98,9 @@
     "glob": "7.2.3",
     "husky": "3.1.0",
     "istanbul": "~0.4.2",
-    "jest": "26.6.3",
-    "jest-circus": "26.6.3",
-    "jest-cli": "26.6.3",
+    "jest": "28.1.2",
```
[the remark this answers is not in the document]
Comment (petersg83) [reply package.json:101]: yes it's planned

File: packages/core/upload/server/services/metrics.js (hunk 7 above)
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:75]: Yes I think it's the best too. I can handle it (I hesitated to prioritize other things but why not doing it)

File: packages/core/strapi/lib/services/cron.js [hunk 9]
```diff
@@ -1,13 +1,18 @@
...
-const { isFunction } = require('lodash/fp');
+const { isFunction, forEach } = require('lodash/fp');
 
 const createCronService = () => {
-  let jobsSpecs = [];
+  let jobsSpecsMap = {};
 
   return {
-    add(tasks = {}) {
+    add(tasks = {}, namespace) {
```
Comment (gu-stav) [inline packages/core/strapi/lib/services/cron.js:10]: I guess this requires to update the documentation? Also: without a default value for `namespace` that would be breaking change, no? Should we add a default value?

File: packages/core/upload/server/__tests__/bootstrap.test.js [hunk 10]
```diff
@@ -35,7 +35,13 @@ describe('Upload plugin bootstrap function', () => {
         },
       },
       plugins: {
-        upload: {},
+        upload: {
+          services: {
+            metrics: {
+              startRegularMetricsUpdate() {},
```
Comment (gu-stav) [inline packages/core/upload/server/__tests__/bootstrap.test.js:41]: I'm a bit confused about the name of the method: what is an `irregular` metrics update? Should we rename this to `startScheduledMetricsUpdates` for example?

File: packages/core/upload/server/services/metrics.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        folderNumber,
+        averageDepth,
+        maxDepth,
+        averageDeviationDepth,
+      };
+    },
+
+    async startRegularMetricsUpdate() {
+      if (running) {
+        throw new Error('Regular metrics updates are already running');
```
Comment (gu-stav) [inline packages/core/upload/server/services/metrics.js:89]: Same here: I think you mean `scheduled` or `Regulary` ...

File: packages/core/strapi/lib/services/cron.js (hunk 9 above)
Comment (petersg83) [reply packages/core/strapi/lib/services/cron.js:10]: It is not a documented feature. So it's alright no?

File: packages/core/upload/server/__tests__/bootstrap.test.js (hunk 10 above)
Comment (petersg83) [reply packages/core/upload/server/__tests__/bootstrap.test.js:41]: scheduled works for me too ! `regular` means to me that it happens at a constant frequency. We have the same work in French. Maybe in english this sense is less used?
Capture d’écran 2022-07-20 à 14 18 22

Comment (derrickmehaffy) [reply packages/core/upload/server/__tests__/bootstrap.test.js:41]: It has the same literal meaning "we now return to our regularly scheduled programming" from old TV but in tech it doesn't carry the same weight. I'd suggest scheduled here also.

File: packages/core/strapi/lib/services/cron.js (hunk 9 above)
Comment (petersg83) [reply packages/core/strapi/lib/services/cron.js:10]: `namespace = common` for instance?

File: packages/core/strapi/lib/Strapi.js [hunk 12]
```diff
@@ -416,7 +413,11 @@ class Strapi {
 
     await this.runLifecyclesFunctions(LIFECYCLES.BOOTSTRAP);
 
-    this.cron.start();
+    if (strapi.config.get('server.cron.enabled', true)) {
+      const cronTasks = this.config.get('server.cron.tasks', {});
+      this.cron.add(cronTasks, 'user');
+      this.cron.start('user');
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/Strapi.js:419]: I don't understand the need for this specific start in the if?

I would conditionally register but always start the service that will just start all registered jobs and if there are none do nothing.

File: packages/core/strapi/lib/services/cron.js [hunk 13]
```diff
@@ -1,13 +1,18 @@
 'use strict';
 
 const { Job } = require('node-schedule');
-const { isFunction } = require('lodash/fp');
+const { isFunction, forEach } = require('lodash/fp');
 
 const createCronService = () => {
-  let jobsSpecs = [];
+  let jobsSpecsMap = {};
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/cron.js:7]: Base on my previous comment I don't think we would need all those changes we could keep this simpler

File: packages/core/upload/server/services/metrics.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      running = true;
+
+      strapi.cron.add(
```
Comment (alexandrebodin) [inline packages/core/upload/server/services/metrics.js:93]: based on my previous comment again, We could just register this in the register of the plugin and never call start on the cron service.

**Review by alexandrebodin — DISMISSED** [review]

**Review by gu-stav — APPROVED** [review]
Two small yakshaving comments that you can ignore if you want: overall this looks very good to me.

File: packages/core/upload/server/services/metrics.js [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+      averageDeviationDepth,
+    };
+  },
+
+  async sendMetrics() {
+    const metrics = await this.computeMetrics();
+    strapi.telemetry.send('didSendUploadPropertiesOnceAWeek', metrics);
+
+    const metricsInfoStored = await getMetricsStoreValue();
+    await setMetricsStoreValue({ ...metricsInfoStored, lastMetricsUpdate: new Date().getTime() });
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:103]: Would `lastWeeklyUpdate` work for you? and `weeklySchedule` instead of `cronSchedule` ?
So we don't have issues if we want to add a daily or monthly update

[the remark this answers is not in the document]
Comment (gu-stav) [reply packages/core/upload/server/services/metrics.js:103]: I'd have called it `updatedAt` because from that you could always derive whether it should be executed: you only store when was the last time it was executed and when you check it, you can make sense out of it whether it's past a weekly or daily update. I'm worried what would happen if you have a `lastWeeklyUpdate` and `lastDailyUpdate` set, because you might execute it twice ... Storing one value might help to reduce that ambiguity?

File: packages/core/upload/server/services/metrics.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+  return defaultTo({}, value);
+};
+const setMetricsStoreValue = value =>
+  strapi.store.set({ type: 'plugin', name: 'upload', key: 'metrics', value });
+
+const addMinutes = (date, offset) => {
+  const newDate = new Date(date);
+  newDate.setMinutes(newDate.getMinutes() + offset);
+  return newDate;
+};
```
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:20]: Yes, it was to avoid installing a lib but why not

File: packages/core/upload/server/services/metrics.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (petersg83) [reply packages/core/upload/server/services/metrics.js:103]: No, I mean if there are different metrics to send daily for instance.
One metric that is sent on monday for instance and one everyday. I need to separate the 2 use cases.

[the remark this answers is not in the document]
Comment (gu-stav) [reply packages/core/upload/server/services/metrics.js]: I see. Let's go for `weeklySchedule` then :)

File: packages/core/upload/server/services/metrics.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+  async ensureWeeklyStoredCronSchedule() {
+    const metricsInfoStored = await getMetricsStoreValue();
+    const { weeklySchedule: currentSchedule, lastWeeklyUpdate } = metricsInfoStored;
+
+    const now = new Date();
+    let weeklySchedule = currentSchedule;
+
+    if (!currentSchedule || !lastWeeklyUpdate || lastWeeklyUpdate + ONE_WEEK < now.getTime()) {
+      weeklySchedule = getWeeklyCronScheduleAt(add(now, { minutes: 5 }));
```
Comment (Marc-Roig) [inline packages/core/upload/server/services/metrics.js:109]: Maybe there are more things planed for the date-fns, but would be an option to create a function like this here?
( date-fns  weights 6.61 MB unpacked )
```js
function addMinutes(date, minutes) {
    return new Date(date.getTime() + minutes*60000);
}
```

Comment (Marc-Roig) [reply packages/core/upload/server/services/metrics.js:109]: Completely missed the comment from gustav above 😄  forget about this

**Review by Convly — APPROVED** [review]
LGTM

File: packages/core/upload/server/services/metrics.js (hunk 17 above)
Comment (gu-stav) [reply packages/core/upload/server/services/metrics.js:109]: @Marc-Roig I really appreciate the comment though :) I agree it is a heavy dependency, but we do have it already installed and using it in many places and therefore it doesn't make much of a difference.
